src/testRunner/unittests/sys/symlinkWatching.ts

```diff
@@ -0,0 +1,805 @@
+                isMacOs,
+                isWindows,
+            );
+        });
+
+        if (isMacOs || isWindows) {
+            describe("recursive watchDirectory using fsEvents", () => {
+                before(() => {
+                    setupRecursiveFsEvents("recursivefsevents");
+                    setupRecursiveFsEvents("recursivefseventssub");
+                    setupRecursiveFsEvents("recursivefseventsparallel");
+                });
+                verifyRecursiveWatchDirectoryUsingFsEvents(
+                    ts.sys,
+                    (dir, recursive, cb) => fs.watch(dir, { persistent: true, recursive }, cb),
+                    `${root}/recursivefsevents`,
+                    `${root}/linkedrecursivefsevents`,
+                    isMacOs,
+                );
+            });
+        }
+
+        function setupRecursiveFsEvents(recursiveName: string) {
+            ts.sys.writeFile(`${root}/${recursiveName}/sub/folder/file.ts`, "export const x = 10;");
+            ts.sys.writeFile(`${root}/${recursiveName}2/sub/folder/file.ts`, "export const x = 10;");
+            withSwallowException(() => fs.symlinkSync(`${root}/${recursiveName}`, `${root}/linked${recursiveName}`, "junction"));
+            withSwallowException(() => fs.symlinkSync(`${root}/${recursiveName}/sub`, `${root}/${recursiveName}/linkedsub`, "junction"));
+            withSwallowException(() => fs.symlinkSync(`${root}/${recursiveName}2`, `${root}/${recursiveName}/linkedsub2`, "junction"));
+        }
+    });
+
+    describe("with virtualFileSystem::", () => {
+        const root = ts.normalizePath("/tests/baselines/symlinks");
+        function getSys(osFlavor?: TestServerHostOsFlavor) {
+            return createWatchedSystem({
+                [`${root}/folder/file.ts`]: "export const x = 10;",
+                [`${root}/linked`]: { symLink: `${root}/folder` },
+            }, { osFlavor });
+        }
+        verifyWatchFile(
+            "watchFile using polling",
+            getSys(),
+            `${root}/folder/file.ts`,
+            `${root}/linked/file.ts`,
+            { watchFile: ts.WatchFileKind.PriorityPollingInterval },
+        );
+        verifyWatchFile(
+            "watchFile using fsEvents",
+            getSys(),
+            `${root}/folder/file.ts`,
+            `${root}/linked/file.ts`,
+            { watchFile: ts.WatchFileKind.UseFsEvents },
+        );
+
+        verifyWatchFile(
+            "watchDirectory using polling",
+            getSys(),
+            `${root}/folder`,
+            `${root}/linked`,
+            { watchFile: ts.WatchFileKind.PriorityPollingInterval },
+            getFileName(),
+        );
+
+        verifyWatchDirectoryUsingFsEvents(
+            getSys(TestServerHostOsFlavor.Windows),
+            (dir, recursive, cb, sys) => sys.fsWatchWorker(dir, recursive, cb),
+            `${root}/folder`,
+            `${root}/linked`,
+            /*isMacOs*/ false,
+            /*isWindows*/ true,

```

Comment:
The "flavor" and the named parameters seem redundant; maybe `verifyWatchDirectoryUsingFsEvents` can just infer the bools from the flavor and construct a sys itself? (And `verifyWatchFile` too?)